### DISCUSSION TITLE
Multiple Instances: Suggestion Plugin

### DISCUSSION THREAD

**Discussion by Rykuno** [ueberdosis/tiptap#1768] [General Questions]
A requirement for utilizing TipTap within our application are multiple suggestion triggers.

Is there any capability for this within TipTap? We tried implementing a solution but it seems one plugin simply overrides the other. 

## Example
Hello `@user` welcome to `$group`

Where `@` mentions a `User`
Where `$` mentions a `Group`

**Comment by philippkuehn** [comment]
That should be possible. could you provide a codesandbox that shows the issue please?

**Reply by GitHubish** [reply]
@philippkuehn I have the impression that the concern corresponds to the issue I have raised here today [#1816](https://github.com/ueberdosis/tiptap/issues/1816). No ?

I have an error when I add the Mention extension twice in the editor. I didn't have this error before.

**Reply by GitHubish** [reply]
Here is an example with the Angular Framework implementation (by @sibiraj-s ) where the problem occurs :  [stackblitz example](https://stackblitz.com/edit/ngx-tiptap-mentions-owzctg?file=src%2Fapp%2Fapp.component.ts)

**Reply by GitHubish** [reply]
Error on my part, I think it does not correspond to my problem. Would setting the suggestion extension to allow multiple characters as a trigger be a good solution? 

For example:
```
Mention.configure({
      HTMLAttributes: {
        class: 'mention',
      },
      suggestion: {
        chars: ['@', '$'],
        allowSpaces: allowSpace,
        allow: ({ editor, range }) => {
```

**Reply by Rykuno** [reply]
Use multiple instances of the Mention within the `useEditor` config. 

For each isntance of mention, generate and use a unique pluginKey.
```js
import { PluginKey } from "prosemirror-state"
export const SuggestionPluginKey = new PluginKey("suggestion1")

const suggestion = {
  char: "u/",
  pluginKey: SuggestionPluginKey,
  items: ({ query }) => {
    return [
      "Lea Thompson",
...
```

**Reply by d8vjork** [reply]
Hey! Had the same issue, and thanks to @Rykuno solution I managed it, but how to solve this warning? It seems related to the name of the node itself but with a `Node.configure` we cannot access that property...

![image](https://user-images.githubusercontent.com/2331052/149356335-4efd573d-eb1e-4b68-9fff-ce7c4045c0a6.png)

**Reply by Rykuno** [reply]
> Hey! Had the same issue, and thanks to @Rykuno solution I managed it, but how to solv

I believe that is solved within where you initialize the editor like below

```jsx
  const editor = useEditor({
    onUpdate({ editor }) {
      setValue(name, editor.getHTML())
    },
    extensions: [
      ...
      Mention.extend({
        name: "userMention",
      }).configure({
        HTMLAttributes: {
          class: styles["mention-user"],
        },
        suggestion: userSuggestions({ refetchUsers }),
      }),
      Mention.extend({
        name: "guildMention",
      }).configure({
        HTMLAttributes: {
          class: styles["mention-guild"],
        },
        suggestion: guildSuggestions({ refetchGuilds }),
      }),
    ],
  })
  ```
  
  So each plugin will have a different `pluginKey` but also a different `name` within the TipTap hook.
  
  Also, for each PluginKey you can provide a unique name as well.
  ```jsx
  const userSuggestions = ({ refetchUsers }: Props) => ({
  char: "u/",
  pluginKey: new PluginKey("userMention"),
  items: async ({ query }: { query: string }) => {
    const { data } = await refetchUsers({
      query,
    })
```

**Reply by moshfiqrony** [reply]
Thanks @Rykuno

**Reply by mrozanski** [reply]
I found this while trying to support the same feature (multiple suggestion triggers).
The suggested solution works but I can't get rid of the warning.
Changing the name seems to be an issue.

The documentation says:

> The same applies to every aspect of an existing extension, except to the name

https://tiptap.dev/docs/editor/guide/custom-extensions#extend-existing-extensions

We have two extensions to the Mention extension, both use a similar suggestion function, they do the same, the difference is in the character that triggers them: `~` and `@`

If I add a name this way, the second one stops working, it triggers the suggestions popup but I don't receive back the selected value, which works without the changed name.
(The warning goes away though)

```
extensions.unshift(
  Mention.extend({ priority: 1000, name: 'mention' }).configure({
    // ...
});
extensions.unshift(
  Mention.extend({ priority: 999, name: 'at-mention' }).configure({
  // ...
})
```

**Reply by verveguy** [reply]
Name cannot be changed. I've resorted to having two copies of the Mention source, each with a different name: set.

**Comment by danmilward** [comment]
> A requirement for utilizing TipTap within our application are multiple suggestion triggers.
> 
> Is there any capability for this within TipTap? We tried implementing a solution but it seems one plugin simply overrides the other.
> ## Example
> 
> Hello `@user` welcome to `$group`
> 
> Where `@` mentions a `User` Where `$` mentions a `Group`

Glad to know we're not the only ones and thanks for the possible fix. 

We need a solution where `@` displays `menu1` and `+` shows `menu2` where menu 2 is a chained feature.

**Comment by arnaugomez** [comment]
Hi @Rykuno, sorry for the delay.

The Mention extension does not support multiple triggers. However, you can achieve this behavior by copying and pasting the Mention extension code to create a new extension. I created this CodeSandbox that modifies the [original Mention demo](https://tiptap.dev/docs/editor/extensions/nodes/mention).

CodeSandbox: https://codesandbox.io/p/sandbox/n5yh22

Gist: https://gist.github.com/arnaugomez/7f83132cb1d9abfb51735ddda9379bb5

This should solve the problem, but it's quite hackey and I don't like it. In the future, I'd like the Mention extension to support different types of mentions, each with its own trigger. I'll try to raise awareness to the team.

**Comment by arnaugomez** [comment]
I created a PR to support multiple triggers in the Mention extension: #6300
